### DISCUSSION TITLE
AWS SDK for Java V2: PutObject corrupts uploaded file when checksumAlgorithm is specified.

### DISCUSSION THREAD

**Discussion by seanmyles2002** [aws/aws-sdk-java-v2#3782] [General]
This is a duplicate of a question I posted on StackExchange a few days ago but since I've had no responses there, I thought I would try this forum. If this is an inappropriate location, I apologize.

I'm trying to use the AWS SDK for Java v2 to upload a file to S3 and have the upload integrity checked using the SHA256 trailing checksum. My code works fine to upload the file if the checksumAlgorithm(ChecksumAlorithm.SHA256) property of PutObjectRequest is not present but when it is included the uploaded file has additional characters at its start and is truncated at the end.

My code is pasted below. My myS3Client is built using the S3Client.builder() method:

    byte[] smallJsonBytes = getObjectFile("UploadTest\\smallFile.txt");
    PutObjectRequest putOb = PutObjectRequest.builder()
          .bucket(credsProvider.getS3Bucketname())
          .key(s3Key)
          **.checksumAlgorithm(ChecksumAlgorithm.SHA256)**
          .build();

    PutObjectResponse response = this.myS3Client.putObject(putOb, RequestBody.fromBytes(smallJsonBytes));

Below is the debug output when the checksumAlgorithm property is included. You can see the AWS SDK is generating a trailing checksum as expected but, as noted already, the uploaded file is corrupted. I've tried specifying the contentLength directly in the PutObjectRequest and also tried using converting smallJsonBytes to a ByteArrayInputStream and specifying the contentLength in the RequestBody both with the same result.

14:56:56.150 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-content-sha256: STREAMING-UNSIGNED-PAYLOAD-TRAILER[\r][\n]"
14:56:56.150 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "X-Amz-Date: 20230216T215655Z[\r][\n]"
14:56:56.150 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-decoded-content-length: 20[\r][\n]"
14:56:56.150 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-sdk-checksum-algorithm: SHA256[\r][\n]"
14:56:56.150 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-trailer: x-amz-checksum-sha256[\r][\n]"
14:56:56.150 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "Content-Length: 99[\r][\n]"
14:56:56.150 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "Connection: Keep-Alive[\r][\n]"
14:56:56.150 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "[\r][\n]"
14:56:56.206 [main] DEBUG org.apache.http.wire - http-outgoing-0 << "HTTP/1.1 100 Continue[\r][\n]"
14:56:56.206 [main] DEBUG org.apache.http.wire - http-outgoing-0 << "[\r][\n]"
14:56:56.211 [main] DEBUG org.apache.http.headers - http-outgoing-0 << HTTP/1.1 100 Continue
14:56:56.211 [main] DEBUG software.amazon.awssdk.core.internal.io.SdkLengthAwareInputStream - Specified InputStream length of 20 has been reached. Returning EOF.
14:56:56.213 [main] DEBUG software.amazon.awssdk.core.internal.io.SdkLengthAwareInputStream - Specified InputStream length of 20 has been reached. Returning EOF.
14:56:56.216 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "14[\r][\n]"
14:56:56.217 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "Nothing much in here[\r][\n]"
14:56:56.217 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "0[\r][\n]"
14:56:56.217 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-checksum-sha256:BthkalBn8SFAnEAFIWr2vtfO0h+TEdX/vOH+iZqKJ4k=[\r][\n]"
14:56:56.217 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "[\r][\n]"
14:56:56.389 [main] DEBUG org.apache.http.wire - http-outgoing-0 << "HTTP/1.1 200 OK[\r][\n]"

**Comment by debora-ito** [comment]
Hi @seanmyles2002 I can't reproduce the issue, setting `ChecksumAlgorithm.SHA256` works fine, the file is not corrupted after the upload.

Does the issue happen too if you upload the file directly via `RequestBody.fromFile()`?
Which version of the Java SDK are you using?

**Reply by seanmyles2002** [reply]
Hi [@debora-ito ](https://github.com/debora-ito),

It's a Maven project and the sdk is included as below:   

` <dependency>
      <groupId>software.amazon.awssdk</groupId>
      <artifactId>s3</artifactId>
      <version>2.19.17</version>
    </dependency>`

When I try to use fromFile() instead, I get the following debug output:


14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "Content-encoding: aws-chunked[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "Content-Type: text/plain[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "Expect: 100-continue[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "User-Agent: aws-sdk-java/2.19.17 Windows_10/10.0 OpenJDK_Client_VM/25.312-b07 Java/1.8.0_312 vendor/Temurin io/sync http/Apache cfg/retry-mode/legacy[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-content-sha256: STREAMING-UNSIGNED-PAYLOAD-TRAILER[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "X-Amz-Date: 20230221T211815Z[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-decoded-content-length: 20[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-sdk-checksum-algorithm: SHA256[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-trailer: x-amz-checksum-sha256[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "Content-Length: 99[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "Connection: Keep-Alive[\r][\n]"
14:18:16.204 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "[\r][\n]"
14:18:16.262 [main] DEBUG org.apache.http.wire - http-outgoing-0 << "HTTP/1.1 100 Continue[\r][\n]"
14:18:16.262 [main] DEBUG org.apache.http.wire - http-outgoing-0 << "[\r][\n]"
14:18:16.267 [main] DEBUG org.apache.http.headers - http-outgoing-0 << HTTP/1.1 100 Continue
14:18:16.270 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "b[\r][\n]"
14:18:16.270 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "Hello World[\r][\n]"
14:18:16.270 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "0[\r][\n]"
14:18:16.270 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "x-amz-checksum-sha256:pZGm1Av0IEBKARczz7exkNYsZb8LzaMrV7J32a2fFG4=[\r][\n]"
14:18:16.270 [main] DEBUG org.apache.http.wire - http-outgoing-0 >> "[\r][\n]"
14:18:46.280 [main] DEBUG org.apache.http.wire - http-outgoing-0 << "[read] I/O error: Read timed out"
14:18:46.280 [main] DEBUG org.apache.http.impl.conn.DefaultManagedHttpClientConnection - http-outgoing-0: Close connection

The error is different and the file upload fails. Also, you can see the additional characters 'b' and '0' being inserted into the upload.

Sean

**Reply by debora-ito** [reply]
> Also, you can see the additional characters 'b' and '0' being inserted into the upload

Those characters are part of the S3 protocol for the `x-amz-trailer` header, not part of the file content. They won't show up in the uploaded file. 

Getting the "Read timeout" is a strange thing in this case though. Do you get "Read timeout" every time you run the test?

**Reply by seanmyles2002** [reply]
If I use fromFile(), then, yes, I get the "Read timed out" message every time.

> Those characters are part of the S3 protocol for the x-amz-trailer header, not part of the file content. They won't show up in the uploaded file.

In the original debug output I included where I was using fromBytes(), you can see this: "14[\r][\n]" : The '14' does appear in the uploaded file (which is also truncated).

**Reply by debora-ito** [reply]
This is weird behavior. Questions to try to narrow down the root cause:
- Are you sending the requests to an S3 endpoint or a third-party solution?
- Are you using any intermediary party, like proxies, that may be changing the request headers?
- Can you share a self-contained sample code and the file you are using to upload, so I can simply run the code and reproduce the issue?

**Reply by seanmyles2002** [reply]
> * Are you sending the requests to an S3 endpoint or a third-party solution?

That's an insightful question: My endpoint is Scality S3 (or whatever it's called these days). I guess there's a possibility that their S3 implementation does not handle the additional checksums.

> Are you using any intermediary party, like proxies, that may be changing the request headers?

I don't think so.

> * Can you share a self-contained sample code and the file you are using to upload, so I can simply run the code and reproduce the issue?

Unfortunately not: The S3 endpoint is not for public consumption. If you have a Scality S3 endpoint you could test against, I could mock my client code.

**Reply by debora-ito** [reply]
> That's an insightful question: My endpoint is Scality S3 (or whatever it's called these days). I guess there's a possibility that their S3 implementation does not handle the additional checksums.

In this case we recommend you reach out to the Scality support team. In this channel we provide help in cases where the requests are sent directly to AWS endpoints, we can't guarantee that the AWS SDK for Java will work with every third-party solution.

**Reply by seanmyles2002** [reply]
Yeah, I'm in the process of doing that now. Thanks for your help so far.

Sean
